app/services/websocket_server.py

```diff
@@ -142,24 +146,30 @@ def format_order(record):
         coin_contract = ERC20Token(record["token_get"])
         base_contract = ERC20Token(record["token_give"])
 
-        available_volume = record["amount_get"] - record["amount_fill"]
+        price = base_contract.denormalize_value(record["amount_give"]) / coin_contract.denormalize_value(record["amount_get"])
+
+        if record["available_volume"] is not None:
+            available_volume = record["available_volume"]
+        else:
+            available_volume = record["amount_get"]
         eth_available_volume = coin_contract.denormalize_value(available_volume)
 
-        filled_base = record["amount_fill"] * record["amount_give"] / record["amount_get"]
-        available_volume_base = record["amount_give"] - filled_base
+        available_volume_base = available_volume * record["amount_give"] / record["amount_get"]
         eth_available_volume_base = base_contract.denormalize_value(available_volume_base)
     else:
         coin_contract = ERC20Token(record["token_give"])
         base_contract = ERC20Token(record["token_get"])
 
-        available_volume = record["amount_give"] - record["amount_fill"]
-        eth_available_volume = coin_contract.denormalize_value(available_volume)
+        price = base_contract.denormalize_value(record["amount_get"]) / coin_contract.denormalize_value(record["amount_give"])
 
-        filled_base = record["amount_fill"] * record["amount_get"] / record["amount_give"]
-        available_volume_base = record["amount_get"] - filled_base
+        if record["available_volume"] is not None:
+            available_volume_base = record["available_volume"]
+        else:
+            available_volume_base = record["amount_get"]
         eth_available_volume_base = base_contract.denormalize_value(available_volume_base)
 
-    price = eth_available_volume_base / eth_available_volume if eth_available_volume > 0 else 0.0
+        available_volume = available_volume_base * record["amount_give"] / record["amount_get"]
```

Comment:
Could use some parentheses in this equation.

Reply:
They'd be around `(available_volume_base * record["amount_give"])`, because multiplying before dividing reduces the chances of an underflow, but I also feel that's more confusing than no parenthesis at all.

Reply:
I personally think adding parenthesis is the way to go. Ultimately up to you, but it makes it more clear to me and very explicitly shows the order of operations.